### PR TITLE
3629 Indices mapping changes before running a new, complete re-index.

### PR DESCRIPTION
**Pull request by albertisfu**
I went through the ES issues to verify we're not missing something related to mapping changes on ES indices before executing the new re index.

### `date_created`

I found this [issue](https://github.com/freelawproject/courtlistener/issues/3109#issuecomment-1828891395) where it's required to add the `date_created` field to the OA index and also it's recommend to add it to all the other indices. 

I added it `date_created` in this PR to all the indices.

### ids as keyword field

ES documentation recommends using the `keyword` field to store integers that do not require range queries, as filters are more performant this way. In  https://github.com/freelawproject/courtlistener/issues/3536#issuecomment-1962134846 I found that changing ID fields to keywords does not show a visible performance improvement. It could be that the performance improvement is too small to be noticeable in a small index like the one used for benchmarks. Besides performance, I mentioned that it might be a good idea to move ID fields (docket_id, recap document id, etc.) to a keyword field or long field, considering that ES `int` field have a maximum value of 2^31 or 2,147,483,648 which enough for now, but thinking about the future, it could be a good idea to move ID fields to a field that supports longer values.

### Content sealed and blocked.

Another change that might require a mapping tweak is related to content that is sealed or blocked. Do we need to hide sealed or blocked content from the search results? Currently, it is possible to find blocked Dockets or sealed RECAP Documents. If we need to hide these from results, we have two options:

- **Remove them from the index:** We'll need to tweak the sweep indexer to remove any content that is blocked or sealed. Also, tweak the indexing signals to avoid indexing content that is blocked or sealed.
- **Index the `blocked` and `sealed` fields:** If we index these fields, we can preserve the documents in the index and hide them from results by including a default filter to avoid retrieving content that is blocked or sealed.

Let me know what you think about it, so we can apply the changes if required.

### CODE REVIEW

File: cl/search/api_serializers.py
Comment (mlissner) [inline cl/search/api_serializers.py:1]: Why exclude `date_created` here?

**Review by mlissner — COMMENTED** [review]
Thanks for this review of outstanding issues. One little question about `date_created`, but the code looks fine.

For the IDs, I don't think we'll hit 2B documents soon, so I'm OK ignoring that for now (sorry future-self). 

For blocked content, we're only trying to keep that content out of *other* search engines, so it's not relevant here. 

For sealed content, I think we're fine because when we remove the content from our DB and Elastic will get automatically updated when that happens, right?

File: cl/search/api_serializers.py
Comment (albertisfu) [reply cl/search/api_serializers.py:1]: I thought this new field might not be useful for users when showing the results via the Search API for OA and Person (documents that are currently using ES in the API), but I can remove the exclusion if necessary.

**Comment by albertisfu** [comment]
About sealed content:

> For sealed content, I think we're fine because when we remove the content from our DB and Elastic will get automatically updated when that happens, right?

Yes, this is true, but I think it applies only in the case where we manually delete the sealed content via the Admin?

For instance, if a `RECAPDocument` is sealed, it is not removed from the database; only its `is_sealed` field is set to True, correct? So, these documents are indexed into ES even though they're marked as sealed and can be found via the search engine and match their content via the snippet.

In the case of `DocketEntries` sealed from RECAP.email, I think we don't have this problem since the entries are not added to the database or ES if they're detected as sealed. However, if we change the approach in the future to add a new `is_sealed` column to the `DocketEntry` model in order to preserve the entry and only avoid showing it, we'd need to do something additional on the ES side to avoid showing or indexing them.

File: cl/search/api_serializers.py
Comment (mlissner) [reply cl/search/api_serializers.py:1]: Yeah, I consider new fields to be backward compatible, so might as well include them.

**Comment by mlissner** [comment]
> if a RECAPDocument is sealed, it is not removed from the database; only its is_sealed field is set to True, correct? 

If you set something as sealed via the admin, it's removed from the DB, S3, and from Internet Archive, so that'd update ES, right? 

Do we set things as sealed in other ways that wouldn't update ES (or Internet Archive or S3)?

File: cl/search/api_serializers.py
Comment (albertisfu) [reply cl/search/api_serializers.py:1]: Great. Now `date_created` is included within the API fields for Judges and OA.

**Comment by albertisfu** [comment]
> If you set something as sealed via the admin, it's removed from the DB, S3, and from Internet Archive, so that'd update ES, right?

I checked the admin method that for sealing content and I found it uses a queryset `update`, which doesn't trigger signals.
We have two options there, use a `save()` method instead or create a custom task that remove sealed content from ES.
 I'll open an issue about it, so new sealed content is removed from ES.
This is not a blocker for starting the re index since it'll take content directly from the DB. 

> Do we set things as sealed in other ways that wouldn't update ES (or Internet Archive or S3)?
The other sealed method is in RECAP.email. Which mark sealed content just setting up `is_sealed` to True. However it's not a problem since in those case there is no PDF available to parse and index.
And Sealed entries are not added at all.

### pacer_case_id field
I just remembered that I noticed we're not indexing the `pacer_case_id` on Dockets or RECAPDocuments, unlike with Oral Arguments. So, I'm wondering if it would be useful to index it for RECAP as well? I think it'll be useful for looking up dockets faster when debugging things.

**Comment by albertisfu** [comment]
Sure, I've filed the issue related to the `seal_documents` action and to also look into other places where `update()` is used.


I've added the `pacer_case_id` field to the RECAP mapping.
However, I verified that for this field, we'll need to update the `recap_vectors` mapping manually. Unlike `date_created`, for which the dynamic mapping creates the right field in the mapping, the `pacer_case_id` field, when created dynamically, results in a multi-field:

```
"pacer_case_id": {
    "type": "text",
    "fields": {
        "keyword": {
            "type": "keyword",
            "ignore_above": 256
        }
    }
}
```

It should be a simple keyword field, since this field is not analyzed for a text query.
```
"pacer_case_id": {
    "type": "keyword"
}
```

**To avoid using unnecessary space, we could just update the index manually before this one is merged.**

```
POST /recap_vectors/_close

PUT /recap_vectors/_mapping
{
   "properties":{
      "absolute_url":{
         "type":"keyword",
         "index":false
      },
      "assignedTo":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "assigned_to_id":{
         "type":"keyword"
      },
      "attachment_number":{
         "type":"integer"
      },
      "attorney":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "attorney_id":{
         "type":"integer"
      },
      "caseName":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "case_name_full":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "cause":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "chapter":{
         "type":"text",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "cites":{
         "type":"integer"
      },
      "court":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "court_citation_string":{
         "type":"text",
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "court_exact":{
         "type":"keyword",
         "index":false
      },
      "court_id":{
         "type":"text",
         "fields":{
            "raw":{
               "type":"keyword"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "dateArgued":{
         "type":"date"
      },
      "dateFiled":{
         "type":"date"
      },
      "dateTerminated":{
         "type":"date"
      },
      "date_created":{
         "type":"date"
      },
      "description":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "docketNumber":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "docket_absolute_url":{
         "type":"keyword",
         "index":false
      },
      "docket_child":{
         "type":"join",
         "eager_global_ordinals":true,
         "relations":{
            "docket":"recap_document"
         }
      },
      "docket_entry_id":{
         "type":"integer"
      },
      "docket_id":{
         "type":"integer"
      },
      "docket_slug":{
         "type":"keyword",
         "index":false
      },
      "document_number":{
         "type":"long"
      },
      "document_type":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "entry_date_filed":{
         "type":"date"
      },
      "entry_number":{
         "type":"long"
      },
      "filepath_local":{
         "type":"keyword",
         "index":false
      },
      "firm":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "firm_id":{
         "type":"integer"
      },
      "id":{
         "type":"integer"
      },
      "is_available":{
         "type":"boolean"
      },
      "jurisdictionType":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "juryDemand":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "pacer_case_id":{
         "type":"keyword"
      },
      "pacer_doc_id":{
         "type":"keyword"
      },
      "page_count":{
         "type":"integer"
      },
      "party":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "party_id":{
         "type":"integer"
      },
      "plain_text":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "referredTo":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "referred_to_id":{
         "type":"keyword"
      },
      "short_description":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "suitNature":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "term_vector":"with_positions_offsets",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "term_vector":"with_positions_offsets",
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      },
      "timestamp":{
         "type":"date"
      },
      "trustee_str":{
         "type":"text",
         "fields":{
            "exact":{
               "type":"text",
               "analyzer":"english_exact",
               "search_analyzer":"search_analyzer_exact"
            }
         },
         "analyzer":"text_en_splitting_cl",
         "search_analyzer":"search_analyzer"
      }
   }
}

POST /recap_vectors/_open
```

**Comment by albertisfu** [comment]
@blancoramiro @mlissner 

According to our plan, before merging this PR, we need to update the Elasticsearch mapping as described in the comment above.

To do this, we need to close the `recap_vectors` index, meaning all RECAP indexing tasks will fail during this time. Hence, we have two options.

We could disable the Dockets and RECAPDocuments indexing signals by adjusting the following settings in K8s:

`ELASTICSEARCH_RECAP_DOCS_SIGNALS_DISABLED = True`
`ELASTICSEARCH_DOCKETS_SIGNALS_DISABLED = True`

By doing so, indexing signals for RECAP will be disabled. We will miss the updates and new documents added while the signals are disabled.

Alternatively, Mike suggested we could stop the celery workers. Following this option, I assume that indexing signals would continue to be enqueued but will wait in the queue until workers are available again. Thus, we won't miss any updates or new documents, correct? If so, I believe this is the best option.

Once the celery tasks are stopped, we should close the `recap_vectors` index, apply the mapping changes, reopen the index, and restore the celery workers.

After everything is ready, we can merge this PR.

Also, before we start with `sweep_indexer`, we need to decide whether to run `cl_remove_non_recap_dockets_from_es` before or simultaneously, considering that the cluster can handle the load.

**Pull request merged**